File: packages/firebase_admob/android/src/main/java/io/flutter/plugins/firebaseadmob/MobileAd.java
// Copyright 2017 The Chromium Authors. All rights reserved.
// Use of this source code is governed by a BSD-style license that can be
// found in the LICENSE file.

package io.flutter.plugins.firebaseadmob;

import android.app.Activity;
import android.util.Log;
import android.util.SparseArray;
import android.view.Gravity;
import android.view.View;
import android.view.ViewGroup;
import android.widget.LinearLayout;
import com.google.android.gms.ads.AdListener;
import com.google.android.gms.ads.AdSize;
import com.google.android.gms.ads.AdView;
import com.google.android.gms.ads.InterstitialAd;
import io.flutter.plugin.common.MethodChannel;
import java.util.HashMap;
import java.util.Map;

abstract class MobileAd extends AdListener {
  private static final String TAG = "flutter";
  private static SparseArray<MobileAd> allAds = new SparseArray<MobileAd>();

  final Activity activity;
  final MethodChannel channel;
  final int id;
  Status status;
  double anchorOffset;
  int anchorType;

  enum Status {
    CREATED,
    LOADING,
    FAILED,
    PENDING, // The ad will be shown when status is changed to LOADED.
    LOADED,
  }

  private MobileAd(int id, Activity activity, MethodChannel channel) {
    this.id = id;
    this.activity = activity;
    this.channel = channel;
    this.status = Status.CREATED;
    this.anchorOffset = 0.0;
    this.anchorType = Gravity.BOTTOM;
    allAds.put(id, this);
  }

  static Banner createBanner(Integer id, AdSize adSize, Activity activity, MethodChannel channel) {
    MobileAd ad = getAdForId(id);
    return (ad != null) ? (Banner) ad : new Banner(id, adSize, activity, channel);
  }

  static Interstitial createInterstitial(Integer id, Activity activity, MethodChannel channel) {
    MobileAd ad = getAdForId(id);
    return (ad != null) ? (Interstitial) ad : new Interstitial(id, activity, channel);
  }

  static MobileAd getAdForId(Integer id) {
    return allAds.get(id);
  }

  Status getStatus() {
    return status;
  }

  abstract void load(String adUnitId, Map<String, Object> targetingInfo);

  abstract void show();

  void dispose() {
    allAds.remove(id);
  }

  private Map<String, Object> argumentsMap(Object... args) {
    Map<String, Object> arguments = new HashMap<String, Object>();
    arguments.put("id", id);
    for (int i = 0; i < args.length; i += 2) arguments.put(args[i].toString(), args[i + 1]);
    return arguments;
  }

  @Override
  public void onAdLoaded() {
    boolean statusWasPending = status == Status.PENDING;
    status = Status.LOADED;
    channel.invokeMethod("onAdLoaded", argumentsMap());
    if (statusWasPending) show();
  }

  @Override
  public void onAdFailedToLoad(int errorCode) {
    Log.w(TAG, "onAdFailedToLoad: " + errorCode);
    status = Status.FAILED;
    channel.invokeMethod("onAdFailedToLoad", argumentsMap("errorCode", errorCode));
  }

  @Override
  public void onAdOpened() {
    channel.invokeMethod("onAdOpened", argumentsMap());
  }

  @Override
  public void onAdClicked() {
    channel.invokeMethod("onAdClicked", argumentsMap());
  }

  @Override
  public void onAdImpression() {
    channel.invokeMethod("onAdImpression", argumentsMap());
  }

  @Override
  public void onAdLeftApplication() {
    channel.invokeMethod("onAdLeftApplication", argumentsMap());
  }

  @Override
  public void onAdClosed() {
    channel.invokeMethod("onAdClosed", argumentsMap());
  }

  static class Banner extends MobileAd {
    private AdView adView;
    private AdSize adSize;

    private Banner(Integer id, AdSize adSize, Activity activity, MethodChannel channel) {
      super(id, activity, channel);
      this.adSize = adSize;
    }

    @Override
    void load(String adUnitId, Map<String, Object> targetingInfo) {
      if (status != Status.CREATED) return;
      status = Status.LOADING;

      adView = new AdView(activity);
      adView.setAdSize(adSize);
      adView.setAdUnitId(adUnitId);
      adView.setAdListener(this);

      AdRequestBuilderFactory factory = new AdRequestBuilderFactory(targetingInfo);
      adView.loadAd(factory.createAdRequestBuilder().build());
    }

    @Override
    void show() {
      if (status == Status.LOADING) {
        status = Status.PENDING;
        return;
      }
      if (status != Status.LOADED) return;

      if (activity.findViewById(id) == null) {
        LinearLayout content = new LinearLayout(activity);
        content.setId(id);
        content.setOrientation(LinearLayout.VERTICAL);
        content.setGravity(anchorType);
        content.addView(adView);
        final float scale = activity.getResources().getDisplayMetrics().density;

        if (anchorType == Gravity.BOTTOM) {
          content.setPadding(0, 0, 0, (int) (anchorOffset * scale));
        } else {
          content.setPadding(0, (int) (anchorOffset * scale), 0, 0);
        }

        activity.addContentView(
            content,
            new ViewGroup.LayoutParams(
                ViewGroup.LayoutParams.MATCH_PARENT, ViewGroup.LayoutParams.MATCH_PARENT));
      }
    }

    @Override
    void dispose() {
      super.dispose();

      View contentView = activity.findViewById(id);
      if (contentView == null || !(contentView.getParent() instanceof ViewGroup)) return;

      adView.destroy();

      ViewGroup contentParent = (ViewGroup) (contentView.getParent());
      contentParent.removeView(contentView);
    }
  }

  static class Interstitial extends MobileAd {
    private InterstitialAd interstitial = null;

    private Interstitial(int id, Activity activity, MethodChannel channel) {
      super(id, activity, channel);
    }

    @Override
    void load(String adUnitId, Map<String, Object> targetingInfo) {
      status = Status.LOADING;

      interstitial = new InterstitialAd(activity);
      interstitial.setAdUnitId(adUnitId);

      interstitial.setAdListener(this);
      AdRequestBuilderFactory factory = new AdRequestBuilderFactory(targetingInfo);
      interstitial.loadAd(factory.createAdRequestBuilder().build());
    }

    @Override
    void show() {
      if (status == Status.LOADING) {
        status = Status.PENDING;
        return;
      }
      interstitial.show();
    }

    // It is not possible to hide/remove/destroy an AdMob interstitial Ad.
  }
}

File: packages/firebase_admob/ios/Classes/MobileAd.m
// Copyright 2017, the Flutter project authors.  Please see the AUTHORS file
// for details. All rights reserved. Use of this source code is governed by a
// BSD-style license that can be found in the LICENSE file.

#import "MobileAd.h"
#import "FirebaseAdmobPlugin.h"

static NSMutableDictionary *allAds = nil;
static NSDictionary *statusToString = nil;

static void logWarning(NSString *format, ...) {
  va_list args;
  va_start(args, format);
  NSString *message = [[NSString alloc] initWithFormat:format arguments:args];
  va_end(args);
  NSLog(@"FirebaseAdMobPlugin <Warning> %@", message);
}

@implementation MobileAd
NSNumber *_mobileAdId;
FlutterMethodChannel *_channel;
MobileAdStatus _status;

+ (void)initialize {
  if (allAds == nil) {
    allAds = [[NSMutableDictionary alloc] init];
  }

  if (statusToString == nil) {
    statusToString = @{
      @(CREATED) : @"CREATED",
      @(LOADING) : @"LOADING",
      @(FAILED) : @"FAILED",
      @(PENDING) : @"PENDING",
      @(LOADED) : @"LOADED"
    };
  }
}

+ (void)configureWithAppId:(NSString *)appId {
  [GADMobileAds configureWithApplicationID:appId];
}

+ (MobileAd *)getAdForId:(NSNumber *)mobileAdId {
  return allAds[mobileAdId];
}

+ (UIViewController *)rootViewController {
  return [UIApplication sharedApplication].delegate.window.rootViewController;
}

- (instancetype)initWithId:(NSNumber *)mobileAdId channel:(FlutterMethodChannel *)channel {
  self = [super init];
  if (self) {
    _mobileAdId = mobileAdId;
    _channel = channel;
    _status = CREATED;
    allAds[mobileAdId] = self;
  }
  return self;
}

- (MobileAdStatus)status {
  return _status;
}

- (void)loadWithUnitId:(NSString *)unitId targetingInfo:(NSDictionary *)targetingInfo {
  // Implemented by the Banner and Interstitial subclasses
}

- (void)show {
  // Implemented by the Banner and Interstitial subclasses
}

- (void)dispose {
  [allAds removeObjectForKey:_mobileAdId];
}

- (NSArray *)targetingInfoArrayForKey:(NSString *)key info:(NSDictionary *)info {
  NSObject *value = info[key];
  if (value == NULL) {
    return nil;
  }
  if (![value isKindOfClass:[NSArray class]]) {
    logWarning(@"targeting info %@: expected an array (MobileAd %@)", key, self);
    return nil;
  }
  return (NSArray *)value;
}

- (NSString *)targetingInfoStringForKey:(NSString *)key info:(NSDictionary *)info {
  NSObject *value = info[key];
  if (value == NULL) {
    return nil;
  }
  if (![value isKindOfClass:[NSString class]]) {
    logWarning(@"targeting info %@: expected a string (MobileAd %@)", key, self);
    return nil;
  }
  NSString *stringValue = (NSString *)value;
  if ([stringValue length] == 0) {
    logWarning(@"targeting info %@: expected a non-empty string (MobileAd %@)", key, self);
    return nil;
  }
  return stringValue;
}

- (NSNumber *)targetingInfoBoolForKey:(NSString *)key info:(NSDictionary *)info {
  NSObject *value = info[key];
  if (value == NULL) {
    return nil;
  }
  if (![value isKindOfClass:[NSNumber class]]) {
    logWarning(@"targeting info %@: expected a boolean, (MobileAd %@)", key, self);
    return nil;
  }
  return (NSNumber *)value;
}

- (GADRequest *)createLoadRequest:(NSDictionary *)targetingInfo {
  GADRequest *request = [GADRequest request];
  if (targetingInfo == nil) {
    return request;
  }

  NSArray *testDevices = [self targetingInfoArrayForKey:@"testDevices" info:targetingInfo];
  if (testDevices != nil) {
    request.testDevices = testDevices;
  }

  NSArray *keywords = [self targetingInfoArrayForKey:@"keywords" info:targetingInfo];
  if (keywords != nil) {
    request.keywords = keywords;
  }

  NSString *contentURL = [self targetingInfoStringForKey:@"contentUrl" info:targetingInfo];
  if (contentURL != nil) {
    request.contentURL = contentURL;
  }

  NSObject *birthday = targetingInfo[@"birthday"];
  if (birthday != NULL) {
    if (![birthday isKindOfClass:[NSNumber class]]) {
      logWarning(@"targeting info birthday: expected a long integer (MobileAd %@)", self);
    } else {
      // Incoming time value is milliseconds since the epoch, NSDate uses
      // seconds.
      request.birthday =
          [NSDate dateWithTimeIntervalSince1970:((NSNumber *)birthday).longValue / 1000.0];
    }
  }

  NSObject *gender = targetingInfo[@"gender"];
  if (gender != NULL) {
    if (![gender isKindOfClass:[NSNumber class]]) {
      logWarning(@"targeting info gender: expected an integer (MobileAd %@)", self);
    } else {
      int genderValue = ((NSNumber *)gender).intValue;
      switch (genderValue) {
        case 0:  // MobileAdGender.unknown
        case 1:  // MobileAdGender.male
        case 2:  // MobileAdGender.female
          request.gender = genderValue;
          break;
        default:
          logWarning(@"targeting info gender: not one of 0, 1, or 2 (MobileAd %@)", self);
      }
    }
  }

  NSNumber *childDirected = [self targetingInfoBoolForKey:@"childDirected" info:targetingInfo];
  if (childDirected != nil) {
    [request tagForChildDirectedTreatment:childDirected.boolValue];
  }

  NSString *requestAgent = [self targetingInfoStringForKey:@"requestAgent" info:targetingInfo];
  if (requestAgent != nil) {
    request.requestAgent = requestAgent;
  }

  return request;
}

- (NSDictionary *)argumentsMap {
  return @{@"id" : _mobileAdId};
}

- (NSString *)description {
  NSString *statusString = (NSString *)statusToString[[NSNumber numberWithInt:_status]];
  return [NSString
      stringWithFormat:@"%@ %@ mobileAdId:%@", super.description, statusString, _mobileAdId];
}
@end

@implementation BannerAd
GADBannerView *_banner;

+ (instancetype)withId:(NSNumber *)mobileAdId channel:(FlutterMethodChannel *)channel {
  MobileAd *ad = [MobileAd getAdForId:mobileAdId];
  return ad != nil ? (BannerAd *)ad : [[BannerAd alloc] initWithId:mobileAdId channel:channel];
}

- (void)loadWithUnitId:(NSString *)unitId targetingInfo:(NSDictionary *)targetingInfo {
  if (_status != CREATED) return;
  _status = LOADING;
  _banner = [[GADBannerView alloc] initWithAdSize:kGADAdSizeBanner];
  _banner.delegate = self;
  _banner.adUnitID = unitId;
  _banner.rootViewController = [MobileAd rootViewController];
  [_banner loadRequest:[self createLoadRequest:targetingInfo]];
}

- (void)show {
  if (_status == LOADING) {
    _status = PENDING;
    return;
  }
  if (_status != LOADED) return;

  UIView *screen = [MobileAd rootViewController].view;
  CGFloat x = screen.frame.size.width / 2 - _banner.frame.size.width / 2;
  CGFloat y = screen.frame.size.height - _banner.frame.size.height;
  _banner.frame = (CGRect){{x, y}, _banner.frame.size};
  [screen addSubview:_banner];
}

- (void)adViewDidReceiveAd:(GADBannerView *)adView {
  bool statusWasPending = _status == PENDING;
  _status = LOADED;
  [_channel invokeMethod:@"onAdLoaded" arguments:[self argumentsMap]];
  if (statusWasPending) [self show];
}

- (void)adView:(GADBannerView *)adView didFailToReceiveAdWithError:(GADRequestError *)error {
  logWarning(@"adView:didFailToReceiveAdWithError: %@ (MobileAd %@)", [error localizedDescription],
             self);
  [_channel invokeMethod:@"onAdFailedToLoad" arguments:[self argumentsMap]];
}

- (void)adViewWillPresentScreen:(GADBannerView *)adView {
  [_channel invokeMethod:@"onAdClicked" arguments:[self argumentsMap]];
}

- (void)adViewWillDismissScreen:(GADBannerView *)adView {
  [_channel invokeMethod:@"onAdImpression" arguments:[self argumentsMap]];
}

- (void)adViewDidDismissScreen:(GADBannerView *)adView {
  [_channel invokeMethod:@"onAdClosed" arguments:[self argumentsMap]];
}

- (void)adViewWillLeaveApplication:(GADBannerView *)adView {
  [_channel invokeMethod:@"onAdLeftApplication" arguments:[self argumentsMap]];
}

- (void)dispose {
  if (_banner.superview) [_banner removeFromSuperview];
  _banner = nil;
  [super dispose];
}

- (NSString *)description {
  return [NSString stringWithFormat:@"%@ for: %@", super.description, _banner];
}
@end

@implementation InterstitialAd
GADInterstitial *_interstitial;

+ (instancetype)withId:(NSNumber *)mobileAdId channel:(FlutterMethodChannel *)channel {
  MobileAd *ad = [MobileAd getAdForId:mobileAdId];
  return ad != nil ? (InterstitialAd *)ad
                   : [[InterstitialAd alloc] initWithId:mobileAdId channel:channel];
}

- (void)loadWithUnitId:(NSString *)unitId targetingInfo:(NSDictionary *)targetingInfo {
  if (_status != CREATED) return;
  _status = LOADING;

  _interstitial = [[GADInterstitial alloc] initWithAdUnitID:unitId];
  _interstitial.delegate = self;
  [_interstitial loadRequest:[self createLoadRequest:targetingInfo]];
}

- (void)show {
  if (_status == LOADING) {
    _status = PENDING;
    return;
  }
  if (_status != LOADED) return;

  [_interstitial presentFromRootViewController:[MobileAd rootViewController]];
}

- (void)interstitialDidReceiveAd:(GADInterstitial *)ad {
  bool statusWasPending = _status == PENDING;
  _status = LOADED;
  [_channel invokeMethod:@"onAdLoaded" arguments:[self argumentsMap]];
  if (statusWasPending) [self show];
}

- (void)interstitial:(GADInterstitial *)ad didFailToReceiveAdWithError:(GADRequestError *)error {
  logWarning(@"interstitial:didFailToReceiveAdWithError: %@ (MobileAd %@)",
             [error localizedDescription], self);
  [_channel invokeMethod:@"onAdFailedToLoad" arguments:[self argumentsMap]];
}

- (void)interstitialWillPresentScreen:(GADInterstitial *)ad {
  [_channel invokeMethod:@"onAdClicked" arguments:[self argumentsMap]];
}

- (void)interstitialWillDismissScreen:(GADInterstitial *)ad {
  [_channel invokeMethod:@"onAdImpression" arguments:[self argumentsMap]];
}

- (void)interstitialDidDismissScreen:(GADInterstitial *)ad {
  [_channel invokeMethod:@"onAdClosed" arguments:[self argumentsMap]];
}

- (void)interstitialWillLeaveApplication:(GADInterstitial *)ad {
  [_channel invokeMethod:@"onAdLeftApplication" arguments:[self argumentsMap]];
}

- (void)dispose {
  // It is not possible to hide/remove/destroy an AdMob interstitial Ad.
  _interstitial = nil;
  [super dispose];
}

- (NSString *)description {
  return [NSString stringWithFormat:@"%@ for: %@", super.description, _interstitial];
}
@end
